src/FsPickler/PicklerGeneration/FSharpTypeGen.fs

```diff
@@ -27,225 +27,232 @@ type internal FsUnionPickler =
-                let picklers = fields |> Array.map (fun f -> resolver.Resolve f.PropertyType)
-                let tags = fields |> Array.mapi (fun i f -> getNormalizedFieldName i f.Name)
-                ctor, fields, tags, picklers)
-
-        let picklerss = caseInfo |> Array.map (fun (_,_,_,picklers) -> picklers)
-
-        let writerDele =
-            DynamicMethod.compileAction4<Pickler [] [], UnionCaseSerializationHelper, WriteState, 'Union> "unionSerializer" (fun picklerss cs writer union ilGen ->
-                let tag = EnvItem<int>(ilGen)
-                let picklers = EnvItem<Pickler []>(ilGen)
-
-                let labels = Array.init caseInfo.Length (fun _ -> ilGen.DefineLabel())
-
-                // read union tag
-                union.Load ()
-                ilGen.EmitCall(OpCodes.Call, tagReaderMethod, null)
-                tag.Store ()
-
-                // select appropriate picklers & store
-                picklerss.Load ()
-                tag.Load ()
-                ilGen.Emit OpCodes.Ldelem_Ref
-                picklers.Store ()
-
-                // serialize union case tag
-                UnionCaseSerializationHelper.InvokeTagWriter cs writer tag ilGen
-
-                // make jump table
-                tag.Load ()
-                ilGen.Emit(OpCodes.Switch, labels)
-
-                // emit cases
-                for i = 0 to caseInfo.Length - 1 do
-                    let label = labels.[i]
-                    let _,fields,tags,_ = caseInfo.[i]
-
-                    ilGen.MarkLabel label
-                    emitSerializeMembers fields tags writer picklers union ilGen
-                    ilGen.Emit OpCodes.Ret
-            )
-
-        let readerDele =
-            DynamicMethod.compileFunc3<Pickler [] [], UnionCaseSerializationHelper, ReadState, 'Union> "unionDeserializer" (fun picklerss cs reader ilGen ->
-
-                let tag = EnvItem<int>(ilGen)
-                let picklers = EnvItem<Pickler []>(ilGen)
-
-                let labels = Array.init caseInfo.Length (fun _ -> ilGen.DefineLabel())
-
-                // resolve union case tag
-                UnionCaseSerializationHelper.InvokeTagReader cs reader ilGen
-                tag.Store()
-
-                // select appropriate picklers & store
-                picklerss.Load ()
-                tag.Load ()
-                ilGen.Emit OpCodes.Ldelem_Ref
-                picklers.Store ()
-
-                // make jump table
-                tag.Load ()
-                ilGen.Emit(OpCodes.Switch, labels)
-
-                // emit cases
-                for i = 0 to caseInfo.Length - 1 do
-                    let label = labels.[i]
-                    let ctor,fields,tags,_ = caseInfo.[i]
-
-                    let ctorParams = fields |> Array.map (fun f -> f.PropertyType)
+        if ty.IsValueType then
+            StructFieldPickler.Create(resolver)
```

Comment:
@varon the pedant in me would say that this forces serialization of all fields for every union case. In practice this means that struct unions have completely different serialization format than class unions. Ideally these should match, however FsPickler is not a library particularly concerned with serialization format.